text/0035-opentelemetry-protocol.md

```diff
@@ -0,0 +1,425 @@
+
+![Request-Response](images/otlp-request-response.png)
+
+_Note: this protocol is concerned with reliability of delivery between one pair of client/server nodes and aims to ensure that no data is lost in-transit between the client and the server. Many telemetry collection systems have intermediary nodes that the data must travel across until reaching the final destination (e.g. application -> agent -> collector -> backend). End-to-end delivery guarantees in such systems is outside of the scope of OTLP. The acknowledgements described in this protocol happen between a single client/server pair and do not span intermediary nodes in multi-hop delivery paths._
+
+#### OTLP over gRPC
+
+For gRPC transport OTLP uses Unary RPC to send export requests and receives responses.
+
+After sending the request the client MAY wait until the response is received from the server. In that case there will be at most only one request in flight that is not yet acknowledged by the server.
+
+![Unary](images/otlp-sequential.png)
+
+Sequential operation is recommended when simplicity of implementation is desirable and when the client and the server are connected via very low-latency network, such as for example when the client is an instrumented application and the server is a OpenTelemetry Service running as a local daemon.
+
+The implementations that need to achieve high throughput SHOULD support concurrent Unary calls to achieve higher throughput. The client SHOULD send new requests without waiting for the response to the earlier sent requests, essentially creating a pipeline of requests that are currently in flight that are not acknowledged.
+
+![Streaming](images/otlp-concurrent.png)
+
+The number of concurrent requests SHOULD be configurable.
+
+The maximum achievable throughput is `max_concurrent_requests * max_request_size / (network_latency + server_response_time)`. For example if the request can contain at most 100 spans, network roundtrip latency is 200ms and server response time is 300 ms, then the maximum achievable throughput with one concurrent request is `100 spans / (200ms+300ms)` or 200 spans per second. It is easy to see that in high latency networks or when the server response time is high to achieve good throughput the requests need to be very big or a lot concurrent requests must be done.
+
+If the client is shutting down (e.g. when the containing process wants to exit) the client will optionally wait until all pending acknowledgements are received or until an implementation specific timeout expires. This ensures reliable delivery of telemetry data. The client implementation SHOULD expose an option to turn on and off the waiting during shutdown.
+
+If the client is unable to deliver a certain request (e.g. a timer expired while waiting for acknowledgements) the client SHOULD record the fact that the data was not delivered.
```

Comment:
I'd suggest to be explicit here. This is retry-able error, correct?

Reply:
It is retryable in the sense that the server did not tell that it is not retryable. However if the client is shutting down there is no way to retry it unless the client is revived from shutdown state (not defined by this doc) or a new client is created.
I am not sure if it is worth touching these topics in this doc.